File: demo-app/app/src/javaApp/java/com/launchkey/android/authenticator/demo/ui/fragment/CustomUnlinkFragment.java
package com.launchkey.android.authenticator.demo.ui.fragment;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.support.annotation.Nullable;

import com.launchkey.android.authenticator.demo.R;
import com.launchkey.android.authenticator.demo.util.Utils;
import com.launchkey.android.authenticator.sdk.SimpleOperationCallback;
import com.launchkey.android.authenticator.sdk.error.BaseError;
import com.launchkey.android.authenticator.sdk.error.CommunicationError;

/**
 * Created by armando on 7/20/16.
 */
public class CustomUnlinkFragment extends BaseDemoFragment {

    private ProgressDialog mUnlinkingDialog;

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        mUnlinkingDialog = new ProgressDialog(getActivity(), R.style.Theme_WhiteLabel_Dialog);
        mUnlinkingDialog.setIndeterminate(true);
        mUnlinkingDialog.setCancelable(false);
        mUnlinkingDialog.setMessage("Custom Unlinking...");

        unlink();
    }

    private void unlink() {

        Utils.show(mUnlinkingDialog);

        getAuthenticatorManager().unlinkCurrentDevice(new SimpleOperationCallback() {

            @Override
            public void onResult(boolean successful, BaseError error, Object o) {

                Utils.dismiss(mUnlinkingDialog);

                if (successful) {
                    if (getActivity() != null && !getActivity().isFinishing()) {
                        getActivity().finish();
                    }
                } else {
                    String httpCode = error instanceof CommunicationError
                            ? String.format("(%s)", error.getCode()) : "";

                    String message = error == null
                            ? "Null error"
                            : String.format("Error %s obj=%s message=%s",
                            httpCode, error.getClass().getSimpleName(), error.getMessage());
                    Utils.alert(getActivity(), "Problem Unlinking", message);
                }
            }
        });
    }
}

File: demo-app/app/src/kotlinApp/java/com/launchkey/android/authenticator/demo/ui/fragment/CustomUnlinkFragment.kt
package com.launchkey.android.authenticator.demo.ui.fragment

import android.app.ProgressDialog
import android.os.Bundle

import com.launchkey.android.authenticator.demo.R
import com.launchkey.android.authenticator.demo.util.Utils
import com.launchkey.android.authenticator.sdk.SimpleOperationCallback
import com.launchkey.android.authenticator.sdk.error.BaseError
import com.launchkey.android.authenticator.sdk.error.CommunicationError

/**
 * Created by armando on 7/20/16.
 */
class CustomUnlinkFragment : BaseDemoFragment() {

    private var mUnlinkingDialog: ProgressDialog? = null

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        mUnlinkingDialog = ProgressDialog(activity, R.style.Theme_WhiteLabel_Dialog)
        mUnlinkingDialog!!.isIndeterminate = true
        mUnlinkingDialog!!.setCancelable(false)
        mUnlinkingDialog!!.setMessage("Custom Unlinking...")

        unlink()
    }

    private fun unlink() {

        Utils.show(mUnlinkingDialog)

        authenticatorManager!!.unlinkCurrentDevice(object : SimpleOperationCallback<Any> {

            override fun onResult(successful: Boolean, error: BaseError?, o: Any?) {

                Utils.dismiss(mUnlinkingDialog)

                if (successful) {
                    if (activity != null && !activity!!.isFinishing) {
                        activity!!.finish()
                    }
                } else {
                    val httpCode = if (error is CommunicationError)
                        String.format("(%s)", error.code)
                    else
                        ""

                    val message = if (error == null)
                        "Null error"
                    else
                        String.format("Error %s obj=%s message=%s",
                                httpCode, error.javaClass.simpleName, error.message)
                    Utils.alert(activity, "Problem Unlinking", message)
                }
            }
        })
    }
}
